File: models/Gpt-3.5-turbo/cpp/code/problems/graph_algorithms/Road_Reparation.cpp

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Road {
    int city1;
    int city2;
    int cost;
};

bool compareRoads(const Road& r1, const Road& r2) {
    return r1.cost < r2.cost;
}

vector<int> parent;

int find(int x) {
    if (parent[x] == x) return x;
    return parent[x] = find(parent[x]);
}

bool unite(int x, int y) {
    int px = find(x);
    int py = find(y);
    if (px == py) return false;
    parent[px] = py;
    return true;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<Road> roads(m);
    for (int i = 0; i < m; i++) {
        cin >> roads[i].city1 >> roads[i].city2 >> roads[i].cost;
    }

    sort(roads.begin(), roads.end(), compareRoads);

    parent.resize(n+1);
    for (int i = 1; i <= n; i++) {
        parent[i] = i;
    }

    vector<Road> minimumSpanningTree;
    int totalCost = 0;

    for (int i = 0; i < m; i++) {
        if (unite(roads[i].city1, roads[i].city2)) {
            minimumSpanningTree.push_back(roads[i]);
            totalCost += roads[i].cost;
        }
        if (minimumSpanningTree.size() == n-1) break;
    }

    if (minimumSpanningTree.size() != n-1) {
        cout << "IMPOSSIBLE" << endl;
    } else {
        cout << totalCost << endl;
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/graph_algorithms/Road_Reparation.py

n, m = map(int, input().split())

roads = []
for _ in range(m):
    a, b, c = map(int, input().split())
    roads.append((a, b, c))

roads.sort(key=lambda x: x[2])

parent = [i for i in range(n+1)]

def find_parent(city):
    if parent[city] != city:
        parent[city] = find_parent(parent[city])
    return parent[city]

def union(city1, city2):
    parent1 = find_parent(city1)
    parent2 = find_parent(city2)
    if parent1 != parent2:
        parent[parent2] = parent1

total_cost = 0
edge_count = 0
for road in roads:
    a, b, c = road
    if find_parent(a) != find_parent(b):
        union(a, b)
        total_cost += c
        edge_count += 1
    if edge_count == n-1:
        break

if edge_count == n-1:
    print(total_cost)
else:
    print("IMPOSSIBLE")
